File: upload_prod/backend/app/models/transport.py
"""
SQLAlchemy models — Phase 10: Transport Management
Tables: vehicles, routes, stops, student_transport, vehicle_maintenance, vehicle_fuel_log
"""
from __future__ import annotations

from datetime import date, datetime, time
from decimal import Decimal
from typing import Optional, TYPE_CHECKING

from sqlalchemy import (
    Boolean, Date, DateTime, ForeignKey, Index, Integer, Numeric, SmallInteger,
    String, Text, Time, UniqueConstraint,
)
from app.models.compat import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base
from app.models.base import TimestampMixin, UUIDPrimaryKeyMixin

if TYPE_CHECKING:
    from app.models.school import School
    from app.models.staff import Staff
    from app.models.students import Student
    from app.models.academic import AcademicYear


class Vehicle(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "vehicles"
    __table_args__ = (
        UniqueConstraint("school_id", "registration_number", name="vehicles_school_reg_ux"),
    )

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    registration_number: Mapped[str] = mapped_column(String(30), nullable=False)
    vehicle_type: Mapped[str] = mapped_column(String(50), nullable=False, default="bus")
    make: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    model: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    year: Mapped[Optional[int]] = mapped_column(SmallInteger, nullable=True)
    capacity: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=40)
    driver_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("staff.id"), nullable=True
    )
    conductor_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("staff.id"), nullable=True
    )
    insurance_expiry: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    fitness_expiry: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    permit_expiry: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    gps_device_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    driver_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    driver_phone: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    driver_license: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    conductor_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    conductor_phone: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    # Additional vehicle details
    color: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    chassis_number: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    odometer_reading: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    vehicle_status: Mapped[str] = mapped_column(String(20), nullable=False, default='active')  # active|on_route|maintenance|breakdown
    # Live GPS state (cached last known position)
    current_latitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    current_longitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    last_gps_update: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    routes: Mapped[list["Route"]] = relationship("Route", back_populates="vehicle", lazy="noload")
    maintenance_logs: Mapped[list["VehicleMaintenance"]] = relationship(
        "VehicleMaintenance", back_populates="vehicle", lazy="noload", cascade="all, delete-orphan"
    )
    fuel_logs: Mapped[list["VehicleFuelLog"]] = relationship(
        "VehicleFuelLog", back_populates="vehicle", lazy="noload", cascade="all, delete-orphan"
    )
    gps_logs: Mapped[list["VehicleGPSLog"]] = relationship(
        "VehicleGPSLog", back_populates="vehicle", lazy="noload", cascade="all, delete-orphan"
    )


class Route(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "routes"
    __table_args__ = (UniqueConstraint("school_id", "name", name="routes_school_name_ux"),)

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    vehicle_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id"), nullable=True
    )
    route_type: Mapped[str] = mapped_column(String(10), nullable=False, default='both')  # morning|evening|both
    # Start/end geography
    starting_point_name: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)
    ending_point_name: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)
    start_latitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    start_longitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    end_latitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    end_longitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    # Schedule — morning (depot→school)
    morning_departure_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)  # bus leaves starting point
    school_arrival_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)     # bus reaches school
    # Schedule — evening (school→home)
    evening_departure_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)  # bus leaves school
    # Legacy / generic times kept for backward compat
    start_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)
    end_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)
    # Distance / ETA
    total_distance_km: Mapped[Optional[Decimal]] = mapped_column(Numeric(6, 2), nullable=True)
    estimated_duration_minutes: Mapped[Optional[int]] = mapped_column(SmallInteger, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    vehicle: Mapped[Optional["Vehicle"]] = relationship("Vehicle", back_populates="routes", lazy="noload")
    stops: Mapped[list["Stop"]] = relationship(
        "Stop", back_populates="route", lazy="noload",
        cascade="all, delete-orphan", order_by="Stop.stop_order",
    )


class Stop(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "stops"
    __table_args__ = (UniqueConstraint("route_id", "stop_order", name="stops_route_order_ux"),)

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    route_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("routes.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    address: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    landmark: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    stop_order: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    pickup_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)   # morning: bus arrives at stop
    drop_time: Mapped[Optional[time]] = mapped_column(Time, nullable=True)     # evening: bus drops at stop
    latitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    longitude: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 7), nullable=True)
    fare: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    route: Mapped["Route"] = relationship("Route", back_populates="stops", lazy="noload")


class StudentTransport(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "student_transport"
    __table_args__ = (
        UniqueConstraint("student_id", "academic_year_id", name="student_transport_student_year_ux"),
    )

    student_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("students.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    route_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("routes.id"), nullable=False
    )
    stop_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stops.id"), nullable=False
    )
    academic_year_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("academic_years.id"), nullable=False
    )
    subscription_type: Mapped[str] = mapped_column(String(20), nullable=False, default="both")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


class VehicleMaintenance(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "vehicle_maintenance"

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    vehicle_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    maintenance_type: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    maintenance_date: Mapped[date] = mapped_column(Date, nullable=False)
    next_due_date: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    cost: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    vendor_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    odometer_reading: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="completed")
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    vehicle: Mapped["Vehicle"] = relationship("Vehicle", back_populates="maintenance_logs", lazy="noload")


class VehicleGPSLog(Base, UUIDPrimaryKeyMixin):
    """Stores individual GPS pings from vehicle GPS devices or driver app."""
    __tablename__ = "vehicle_gps_logs"
    __table_args__ = (
        Index("ix_vehicle_gps_logs_vehicle_recorded", "vehicle_id", "recorded_at"),
    )

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    vehicle_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id", ondelete="CASCADE"), nullable=False, index=True
    )
    latitude: Mapped[Decimal] = mapped_column(Numeric(10, 7), nullable=False)
    longitude: Mapped[Decimal] = mapped_column(Numeric(10, 7), nullable=False)
    speed_kmph: Mapped[Optional[Decimal]] = mapped_column(Numeric(5, 2), nullable=True)
    heading_degrees: Mapped[Optional[int]] = mapped_column(SmallInteger, nullable=True)
    altitude_meters: Mapped[Optional[Decimal]] = mapped_column(Numeric(8, 2), nullable=True)
    accuracy_meters: Mapped[Optional[Decimal]] = mapped_column(Numeric(6, 2), nullable=True)
    recorded_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    source: Mapped[str] = mapped_column(String(20), nullable=False, default='driver_app')  # device|driver_app|manual
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    vehicle: Mapped["Vehicle"] = relationship("Vehicle", back_populates="gps_logs", lazy="noload")


class VehicleFuelLog(Base, UUIDPrimaryKeyMixin):
    __tablename__ = "vehicle_fuel_log"

    school_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    vehicle_id: Mapped[str] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id", ondelete="NO ACTION"), nullable=False, index=True
    )
    fuel_date: Mapped[date] = mapped_column(Date, nullable=False)
    fuel_quantity: Mapped[Optional[Decimal]] = mapped_column(Numeric(8, 2), nullable=True)
    fuel_cost: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    odometer_reading: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    pump_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)

    vehicle: Mapped["Vehicle"] = relationship("Vehicle", back_populates="fuel_logs", lazy="noload")

File: new_upload/frontend/src/api/transport.ts
import apiClient from './axios';

export type VehicleStatus = 'active' | 'on_route' | 'maintenance' | 'breakdown';
export type RouteType = 'morning' | 'evening' | 'both';
export type SubscriptionType = 'both' | 'pickup' | 'drop';

export interface Vehicle {
  id: string;
  school_id: string;
  registration_number: string;
  vehicle_type: string;
  make?: string;
  model?: string;
  year?: number;
  capacity: number;
  color?: string;
  chassis_number?: string;
  odometer_reading?: number;
  vehicle_status: VehicleStatus;
  // Driver
  driver_id?: string;
  driver_name?: string;
  driver_phone?: string;
  driver_license?: string;
  // Conductor
  conductor_id?: string;
  conductor_name?: string;
  conductor_phone?: string;
  // Documents
  insurance_expiry?: string;
  fitness_expiry?: string;
  permit_expiry?: string;
  // GPS device
  gps_device_id?: string;
  // Live GPS (cached)
  current_latitude?: number;
  current_longitude?: number;
  last_gps_update?: string;
  is_active: boolean;
}

export interface Stop {
  id: string;
  route_id: string;
  name: string;
  address?: string;
  landmark?: string;
  stop_order: number;
  pickup_time?: string;   // morning: when bus arrives to pick up
  drop_time?: string;     // evening: when bus arrives to drop off
  latitude?: number;
  longitude?: number;
  fare: number;
}

export interface Route {
  id: string;
  school_id: string;
  name: string;
  description?: string;
  vehicle_id?: string;
  route_type: RouteType;
  // Starting point (origin / depot)
  starting_point_name?: string;
  start_latitude?: number;
  start_longitude?: number;
  // Ending point (school or last drop)
  ending_point_name?: string;
  end_latitude?: number;
  end_longitude?: number;
  // Morning schedule
  morning_departure_time?: string;   // bus leaves starting point
  school_arrival_time?: string;      // bus reaches school
  // Evening schedule
  evening_departure_time?: string;   // bus leaves school for drops
  // Distance / ETA
  total_distance_km?: number;
  estimated_duration_minutes?: number;
  // Legacy
  start_time?: string;
  end_time?: string;
  is_active: boolean;
  stops: Stop[];
}

export interface StudentTransport {
  id: string;
  student_id: string;
  school_id: string;
  route_id: string;
  stop_id: string;
  academic_year_id: string;
  subscription_type: SubscriptionType;
  is_active: boolean;
}

export interface VehicleMaintenance {
  id: string;
  vehicle_id: string;
  maintenance_type: string;
  description?: string;
  maintenance_date: string;
  next_due_date?: string;
  cost: number;
  vendor_name?: string;
  odometer_reading?: number;
  status: string;
  notes?: string;
}

export interface FuelLog {
  id: string;
  vehicle_id: string;
  fuel_date: string;
  fuel_quantity?: number;
  fuel_cost: number;
  odometer_reading?: number;
  pump_name?: string;
}

export interface GPSLog {
  id: string;
  vehicle_id: string;
  latitude: number;
  longitude: number;
  speed_kmph?: number;
  heading_degrees?: number;
  altitude_meters?: number;
  accuracy_meters?: number;
  recorded_at: string;
  source: string;
  created_at: string;
}

export interface GPSUpdateRequest {
  latitude: number;
  longitude: number;
  speed_kmph?: number;
  heading_degrees?: number;
  altitude_meters?: number;
  accuracy_meters?: number;
  recorded_at?: string;
  source: 'device' | 'driver_app' | 'manual';
}

export interface VehicleLocation {
  vehicle_id: string;
  registration_number: string;
  vehicle_type: string;
  driver_name?: string;
  driver_phone?: string;
  vehicle_status: VehicleStatus;
  current_latitude?: number;
  current_longitude?: number;
  last_gps_update?: string;
  is_active: boolean;
}

export const transportApi = {
  // ── Vehicles ──────────────────────────────────────────────────────────────
  listVehicles: (activeOnly = true) =>
    apiClient.get<Vehicle[]>('/transport/vehicles', { params: { active_only: activeOnly } }),

  getVehicle: (id: string) =>
    apiClient.get<Vehicle>(`/transport/vehicles/${id}`),

  createVehicle: (data: Partial<Vehicle>) =>
    apiClient.post<Vehicle>('/transport/vehicles', data),

  updateVehicle: (id: string, data: Partial<Vehicle>) =>
    apiClient.put<Vehicle>(`/transport/vehicles/${id}`, data),

  deleteVehicle: (id: string) =>
    apiClient.delete(`/transport/vehicles/${id}`),

  getExpiryAlerts: (days = 30) =>
    apiClient.get<object[]>('/transport/vehicles/expiry-alerts', { params: { days } }),

  // ── GPS Tracking ──────────────────────────────────────────────────────────
  recordGPS: (vehicleId: string, data: GPSUpdateRequest) =>
    apiClient.post<GPSLog>(`/transport/vehicles/${vehicleId}/gps`, data),

  getGPSHistory: (vehicleId: string, limit = 100) =>
    apiClient.get<GPSLog[]>(`/transport/vehicles/${vehicleId}/gps`, { params: { limit } }),

  getFleetLocations: () =>
    apiClient.get<VehicleLocation[]>('/transport/fleet/locations'),

  // ── Maintenance ────────────────────────────────────────────────────────────
  listMaintenance: (vehicleId: string) =>
    apiClient.get<VehicleMaintenance[]>(`/transport/vehicles/${vehicleId}/maintenance`),

  addMaintenance: (vehicleId: string, data: Partial<VehicleMaintenance>) =>
    apiClient.post<VehicleMaintenance>(`/transport/vehicles/${vehicleId}/maintenance`, data),

  // ── Fuel Logs ──────────────────────────────────────────────────────────────
  listFuelLogs: (vehicleId: string) =>
    apiClient.get<FuelLog[]>(`/transport/vehicles/${vehicleId}/fuel-logs`),

  addFuelLog: (vehicleId: string, data: Partial<FuelLog>) =>
    apiClient.post<FuelLog>(`/transport/vehicles/${vehicleId}/fuel-logs`, data),

  // ── Routes ─────────────────────────────────────────────────────────────────
  listRoutes: () =>
    apiClient.get<Route[]>('/transport/routes'),

  getRoute: (id: string) =>
    apiClient.get<Route>(`/transport/routes/${id}`),

  createRoute: (data: Partial<Route>) =>
    apiClient.post<Route>('/transport/routes', data),

  updateRoute: (id: string, data: Partial<Route>) =>
    apiClient.put<Route>(`/transport/routes/${id}`, data),

  deleteRoute: (id: string) =>
    apiClient.delete(`/transport/routes/${id}`),

  // ── Stops ──────────────────────────────────────────────────────────────────
  listStops: (routeId: string) =>
    apiClient.get<Stop[]>(`/transport/routes/${routeId}/stops`),

  addStop: (routeId: string, data: Partial<Stop>) =>
    apiClient.post<Stop>(`/transport/routes/${routeId}/stops`, data),

  updateStop: (routeId: string, stopId: string, data: Partial<Stop>) =>
    apiClient.put<Stop>(`/transport/routes/${routeId}/stops/${stopId}`, data),

  deleteStop: (routeId: string, stopId: string) =>
    apiClient.delete(`/transport/routes/${routeId}/stops/${stopId}`),

  // ── Student Assignments ───────────────────────────────────────────────────
  listStudentTransport: (routeId?: string) =>
    apiClient.get<StudentTransport[]>('/transport/students', {
      params: routeId ? { route_id: routeId } : {},
    }),

  assignStudent: (data: Partial<StudentTransport>) =>
    apiClient.post<StudentTransport>('/transport/students', data),

  removeStudentTransport: (id: string) =>
    apiClient.delete(`/transport/students/${id}`),
};
